Additional native query tests for unnest datasource

somu-imply opened the pull request:
Adds the tests and changes for native tests for unnest in Scan, TopN, GroupBy etc.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestTopNQueryRunnerTest.java:
```diff
@@ -0,0 +1,385 @@
+ * under the License.
+ */
+
+package org.apache.druid.query;
```
nit: if this was in `org.apache.druid.query.topn` with all the other topn query tests then wouldn't need change method visibility in `PooledTopNAlgorithm`

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,705 @@
+ * under the License.
+ */
+
+package org.apache.druid.query;
```
nit: imo this should live in `org.apache.druid.query.groupby`

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestScanQueryRunnerTest.java:
```diff
@@ -0,0 +1,530 @@
+ * under the License.
+ */
+
+package org.apache.druid.query;
```
nit: this should live in `org.apache.druid.query.scan`

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,705 @@
+@RunWith(Parameterized.class)
+public class UnnestGroupByQueryRunnerTest extends InitializedNullHandlingTest
+{
+  public static final ObjectMapper DEFAULT_MAPPER = TestHelper.makeSmileMapper();
+  public static final DruidProcessingConfig DEFAULT_PROCESSING_CONFIG = new DruidProcessingConfig()
```
since these and some other stuff are public in the standard `GroupByQueryRunnerTest` I wonder if we should just use those instead of redefining here

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,705 @@
+  @Test
+  public void testGroupByOnUnnestedColumn()
+  {
+    // Cannot vectorize due to extraction dimension spec.
```
nit: this comment isn't true

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,705 @@
+    );
+
+    Iterable<ResultRow> results = GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
+    TestHelper.assertExpectedObjects(expectedResults, results, "hroupBy-on-unnested-column");
```
nit

```suggestion
    TestHelper.assertExpectedObjects(expectedResults, results, "groupBy-on-unnested-column");
```

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,705 @@
+    Iterable<ResultRow> results = GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
+    TestHelper.assertExpectedObjects(expectedResults, results, "hroupBy-on-unnested-column");
+  }
+
```
would be nice to also have a test with a virtual column, maybe both `mv_to_array` and making a composite array from multiple columns with `array` would be interesting

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestScanQueryRunnerTest.java:
```diff
@@ -0,0 +1,530 @@
+  }
+
+  @Test
+  public void testUnnestRunner()
```
nit: `testScan`?

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestScanQueryRunnerTest.java:
```diff
@@ -0,0 +1,530 @@
+  }
+
+  @Test
+  public void testUnnestRunnerVirtualColumns()
```
i think would also be interesting to have a test on a composite array column using `array` function to create an array from multiple columns (in addition to `mv_to_array` test)

clintropolis commented on processing/src/test/java/org/apache/druid/query/UnnestTopNQueryRunnerTest.java:
```diff
@@ -0,0 +1,385 @@
+  }
+
+  @Test
+  public void testTopNStringVirtualColumnUnnest()
```
similar comment about virtual column created from multiple columns using `array` in addition to `mv_to_array` test

clintropolis reviewed (commented):
thanks for adding more tests, just some minor comments :rocket:

imply-cheddar commented on processing/src/test/java/org/apache/druid/query/QueryRunnerTestHelper.java:
```diff
@@ -467,6 +478,87 @@ public String toString()
     };
   }
 
+  public static <T, QueryType extends Query<T>> List<QueryRunner<T>> makeUnnestQueryRunners(
+      QueryRunnerFactory<T, QueryType> factory,
+      String dimensionToUnnest,
+      String outputColumn,
+      LinkedHashSet<String> allowSet
+  )
+  {
+    final IncrementalIndex rtIndex = TestIndex.getIncrementalTestIndex();
+    final IncrementalIndex noRollupRtIndex = TestIndex.getNoRollupIncrementalTestIndex();
+    final QueryableIndex mMappedTestIndex = TestIndex.getMMappedTestIndex();
+    final QueryableIndex noRollupMMappedTestIndex = TestIndex.getNoRollupMMappedTestIndex();
+    final QueryableIndex mergedRealtimeIndex = TestIndex.mergedRealtimeIndex();
+    final QueryableIndex frontCodedMappedTestIndex = TestIndex.getFrontCodedMMappedTestIndex();
```
I'm confused as to why you have to do this?  The Unnest is defined as part of the query, not as part of the query runner, you should be able to just send the query along and it will attach the unnest on the segment via the segment mapper functions, no?

imply-cheddar commented on processing/src/test/java/org/apache/druid/query/UnnestSegment.java:
```diff
@@ -0,0 +1,93 @@
+  private final String renamedOutputDimension;
+  private final LinkedHashSet<String> allowSet;
+
+  public UnnestSegment(Segment baseSegment, String dimension, String outputName, LinkedHashSet<String> allowList)
+  {
+    this.baseSegment = baseSegment;
+    this.dimension = dimension;
+    this.renamedOutputDimension = outputName;
+    this.allowSet = allowList;
+  }
```
I don't believe this should actually be required.  The segment mapping fns should be doing this as part of query processing.

imply-cheddar commented on processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,833 @@
+        .build();
+
+    List<ResultRow> expectedResults = Arrays.asList(
+        makeRow(
+            query,
+            "2011-04-01",
+            "alias",
+            "automotive",
+            "rows",
+            2L,
+            "idx",
+            270L
+        ),
```
style nit, I don't think these are actually wide enough to require each of the arguments being on independent lines.  The test would read a lot easier if each of the `makeRow()` calls was its own line.

imply-cheddar commented on processing/src/test/java/org/apache/druid/query/scan/UnnestScanQueryRunnerTest.java:
```diff
@@ -0,0 +1,606 @@
+                  Map<String, Object> event = new HashMap<>();
+                  String[] values1 = input.split("\\t");
+                  for (int i = 0; i < dimSpecs.length; i++) {
+                    if (dimSpecs[i] == null || i >= dimSpecs.length) {
```
this CodeQL thing is likely legitimate

imply-cheddar reviewed (commented):
I'd like to understand better why you had to pipe things down into the QueryRunner.  You shouldn't've needed to do that, it makes me think that these tests aren't running anything that actually does the segment mapping function...

somu-imply commented:
Migrated towards using the segmentMapFunction along with `SegmentReference` to run the queries. Removed the earlier `UnnestSegment` part. Added scan, groupBy and  topN queries with `array` and `mv_to_array` cases for virtual columns along with other regular test cases. Please review

imply-cheddar commented on .idea/misc.xml:
```diff
@@ -84,7 +87,10 @@
     <output url="file://$PROJECT_DIR$/classes" />
   </component>
-</project>
+  <component name="SuppressKotlinCodeStyleNotification">
+    <option name="disableForAll" value="true" />
+  </component>
+</project>
```
Changes to this file appear to still be checked in?  We cannot merge while this is part of the PR.

somu-imply replied:
Let's merge #13576 and then I'll merge the master with this branch and take care of this one. The UnnestColumnValueSelectorCursor is updated in that PR

somu-imply replied:
This is corrected now

somu-imply commented on processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java:
```diff
@@ -0,0 +1,799 @@
+  public ExpectedException expectedException = ExpectedException.none();
+
+  public UnnestGroupByQueryRunnerTest(
+      String testName,
```
addressed

somu-imply commented:
Updated to remove duplicated code, unused exception and moved to `Numbers.tryParse` to avoid the possible issues CodeQL was pointing to
